### PR TITLE
Issue #52 optimize model classes

### LINKED ISSUE #52: Optimize model classes

### PR DESCRIPTION
**Pull request by crookse**
Closes #52 

My plan for implementing the following syntax is to give users an easy and semantic way to create models as well as getting a JSON representation of the model. I think all of the models we create in this realworld application should follow this syntax, but give me your opinions please. @Guergeiro , @ebebbington .

## Users

Creating users:

```typescript
import UserModel from "path/to/user_model.ts";

const user: UserModel = new UserModel(
  username: string,
  password: string,
  email: string,
  bio: string // optional
  image: string // optional
);

await user.save(); // Can also be assigned to a variable. This returns the UserModel object back.
```

Updating users:

```typescript
import UserModel from "path/to/user_model.ts";

let user: UserModel = UserModel.getUserById(id: numer);
user.username = "newUsername";
user.password = bcrypt.hash("newPassword");

user: UserModel = await user.save(); // Returns the updated UserModel object from the db
```

Deleting users:

```typescript
import UserModel from "path/to/user_model.ts";

const user: UserModel = UserModel.getUserById(id: numer);
const result: boolean = await user.delete(); // Returns true or false based on the result
```

Getting users:

```typescript
import UserModel from "path/to/user_model.ts";

let user: UserModel;
user = UserModel.getUserById(id: numer);
user = UserModel.getUserByUsername(username: string);
user = UserModel.getUserByEmail(email: string);
```

Converting a user model to an entity (a JSON object):

```typescript
import { UserEntity, UserModel } from "path/to/user_model.ts";

const user: UserModel = UserModel.getUserById(id: numer);
const entity: UserEntity = user.toEntity();
```

## Sessions

Syntax for creating sessions:

```typescript
import SessionModel from "path/to/session_model.ts";

const session: SessionModel = new SessionModel(
  sessionOne: string,
  sessionTwo: string,
  userId: number
);

await session.save(); // Can also be assigned to a variable. This returns the SessionModel object back.
```

Getting sessions:

```typescript
import SessionModel from "path/to/user_model.ts";

const session: SessionModel = SessionModel.getUserSession(sessionOneVal: string, sessionTwoVal: string);
```

## Validation Service

```typescript
import ValidationService from "path/to/validation_service.ts";

ValidationService.decodeInput(input: string) // for decoding strings
ValidationService.isEmail(email: string) // for checking if an email is a valid email address
ValidationService.isEmailUnique(email: string) // for checking if an email is not already in the db
ValidationService.isPasswordCorrect(passwordOne: string, passwordTwo: string) // for matching two passwords
ValidationService.isPasswordStrong(password: string) // for checking if a password has 8 characters, 1 number, 1 uppercase letter, and 1 lowercase letter
```

### CODE REVIEW

File: src/resources/users_resource.ts [hunk 1]
```diff
@@ -21,65 +19,91 @@ class RegisterResource extends Drash.Http.Resource {
...
 
-        return this.response;
+    protected errorResponse(message: string): Drash.Http.Response {
+      this.response.status_code = 422;
+      this.response.body = {
+        errors: {
+          body: message
+        }
+      };
+      return this.response;
```
[the remark this answers is not in the document]
Comment (crookse) [reply src/resources/users_resource.ts:106]: Yea that's a good idea. I'll implement that with the error response code

**Comment by ebebbington** [comment]
I get the complete opposite :S I get these errors in my console when logging in / signing up:
```
[Vue warn]: Error in render: "TypeError: Cannot read property 'username' of undefined"

found in

---> <Header> at vue/components/Header.vue
       <App> at vue/components/App.vue
         <Root>
warn @ vue.js:649
logError @ vue.js:1923
globalHandleError @ vue.js:1918
handleError @ vue.js:1878
Vue._render @ vue.js:3516
updateComponent @ vue.js:4039
get @ vue.js:4450
run @ vue.js:4536
flushSchedulerQueue @ vue.js:4284
eval @ vue.js:2024
flushCallbacks @ vue.js:1946
Promise.then (async)
timerFunc @ vue.js:1974
nextTick @ vue.js:2035
queueWatcher @ vue.js:4380
update @ vue.js:4525
notify @ vue.js:766
reactiveSetter @ vue.js:1089
setUser @ mutations.js:16
wrappedMutationHandler @ vuex.esm.js:833
commitIterator @ vuex.esm.js:428
eval @ vuex.esm.js:427
_withCommit @ vuex.esm.js:606
commit @ vuex.esm.js:426
boundCommit @ vuex.esm.js:365
setUser @ actions.js:130
wrappedActionHandler @ vuex.esm.js:840
dispatch @ vuex.esm.js:478
boundDispatch @ vuex.esm.js:361
eval @ actions.js:29
Promise.then (async)
checkIfUserIsAuthenticated @ actions.js:27
wrappedActionHandler @ vuex.esm.js:840
dispatch @ vuex.esm.js:478
boundDispatch @ vuex.esm.js:361
eval @ _app.js:145
iterator @ vue-router.esm.js:2094
step @ vue-router.esm.js:1840
step @ vue-router.esm.js:1844
runQueue @ vue-router.esm.js:1849
confirmTransition @ vue-router.esm.js:2118
transitionTo @ vue-router.esm.js:2020
push @ vue-router.esm.js:2408
eval @ vue-router.esm.js:2711
push @ vue-router.esm.js:2710
eval @ Register.vue?./node_modules/babel-loader/lib!./node_modules/vue-loader/lib??vue-loader-options:87
Promise.then (async)
onSubmit @ Register.vue?./node_modules/babel-loader/lib!./node_modules/vue-loader/lib??vue-loader-options:78
submit @ Register.vue?./node_modules/vue-loader/lib/loaders/templateLoader.js??vue-loader-options!./node_modules/vue-loader/lib??vue-loader-options:44
invokeWithErrorHandling @ vue.js:1888
invoker @ vue.js:2219
original._wrapper @ vue.js:7510
vue.js:1928 TypeError: Cannot read property 'username' of undefined
    at Proxy.render (Header.vue?./node_modules/vue-loader/lib/loaders/templateLoader.js??vue-loader-options!./node_modules/vue-loader/lib??vue-loader-options:153)
    at VueComponent.Vue._render (vue.js:3514)
    at VueComponent.updateComponent (vue.js:4039)
    at Watcher.get (vue.js:4450)
    at Watcher.run (vue.js:4536)
    at flushSchedulerQueue (vue.js:4284)
    at Array.eval (vue.js:2024)
    at flushCallbacks (vue.js:1946)
```

Also, there's definitely a typo somewhere :P i assume it's the session two cookie, but one is being saveed as `sessionTne` after login

**Comment by crookse** [comment]
Ok let me take this back to the drawing board lol. Obviously there are mutliple issues.

**Comment by ebebbington** [comment]
Doesn't _look_ like it'll take too much work to be fair

**Comment by crookse** [comment]
@ebebbington @Guergeiro ,

when you guys tested this, did you rebuild your webpack bundles? i can't reproduce the issues you guys are seeing. i rebuilt my env too

**Comment by ebebbington** [comment]
Actually the session cookies must have been from a previous day - no session cookies are being stored on the client for me

**Comment by ebebbington** [comment]
> @ebebbington @Guergeiro ,
> 
> when you guys tested this, did you rebuild your webpack bundles? i can't reproduce the issues you guys are seeing. i rebuilt my env too

Something weird just happened... I ran `npm run webpack` and i can log in fine now! Lets just put it down to one of those mysterious things, and totally not me forgetting to re-run webpack...

**Comment by ebebbington** [comment]
Same as Breno, really good job on it, it's a really nice approach to handling all the logic around this. And evidently I wouldn't have been able to come up with it :P

Just waiting on the change to the error response method, but to me it's already good go

**Comment by crookse** [comment]
Added the error response stuff in a base resource class: https://github.com/drashland/deno-drash-realworld-example-app/pull/53/commits/2c97445e196522f14eb7b3009703a69c7f268c3a

**Pull request merged**